File: python/src/token_validator/resource.py
"""OAuth 2.0/2.1 protected-resource conventions on top of :class:`AccessTokenValidator`.

A ``ProtectedResource`` is the resource-server side of any OAuth-protected HTTP service — an MCP server
(OAuth 2.1 resource server per the MCP authorization spec), an A2A agent, or a plain REST API. It adds the
transport conventions the token validator itself doesn't cover:

* RFC 9728 Protected Resource Metadata (``/.well-known/oauth-protected-resource``) advertising the
  authorization server(s) — what an MCP server MUST expose so clients can discover where to get a token;
* RFC 6750 bearer-token extraction and the ``WWW-Authenticate`` challenge on a 401;
* a request guard that binds the token's audience to this resource (RFC 8707) and maps failures to the
  right status (401 vs 403).

Protocol-neutral: nothing here is MCP-only.
"""
from __future__ import annotations

from urllib.parse import urlsplit, urlunsplit

from . import errors
from .result import ValidationResult

BEARER = "bearer"
_WELL_KNOWN = "/.well-known/oauth-protected-resource"


def bearer_token(authorization_header):
    """Extract the token from an ``Authorization: Bearer <token>`` header value, or ``None`` if the header
    is absent, blank, or not a Bearer credential."""
    if not authorization_header:
        return None
    parts = authorization_header.split(None, 1)
    if len(parts) == 2 and parts[0].lower() == BEARER and parts[1].strip():
        return parts[1].strip()
    return None


class AuthDecision:
    """The outcome of guarding a request. Truthy when authorized. On failure it carries the HTTP ``status``
    (401 or 403) and the ``www_authenticate`` header value to return."""

    def __init__(self, authorized, *, result=None, status=200, www_authenticate=None,
                 error=None, error_description=None):
        self.authorized = authorized
        self.result = result
        self.status = status
        self.www_authenticate = www_authenticate
        self.error = error
        self.error_description = error_description

    def __bool__(self):
        return self.authorized


# validator error code -> (HTTP status, RFC 6750 error). insufficient_scope is the only 403.
def _http_error(error):
    if error == errors.INSUFFICIENT_SCOPE:
        return 403, "insufficient_scope"
    return 401, "invalid_token"


class ProtectedResource:
    """An OAuth-protected resource server (e.g. an MCP server or A2A agent).

    ``validator`` should be an :class:`AccessTokenValidator` configured so that ``resource`` is an accepted
    audience — that binds incoming tokens to this resource (RFC 8707), which MCP servers MUST enforce.
    """

    def __init__(self, resource, authorization_servers, validator, scopes_supported=None):
        self.resource = resource
        self.authorization_servers = list(authorization_servers or [])
        self.validator = validator
        self.scopes_supported = list(scopes_supported or [])

    def metadata(self):
        """RFC 9728 protected-resource metadata. Serve this JSON at :meth:`metadata_path`."""
        md = {
            "resource": self.resource,
            "authorization_servers": self.authorization_servers,
            "bearer_methods_supported": ["header"],
        }
        if self.scopes_supported:
            md["scopes_supported"] = self.scopes_supported
        return md

    def metadata_path(self):
        """The path to serve the metadata at, per RFC 9728 §3: ``/.well-known/oauth-protected-resource``
        with the resource's path (if any) appended."""
        path = urlsplit(self.resource).path
        return _WELL_KNOWN + path if path and path != "/" else _WELL_KNOWN

    def metadata_url(self):
        """The absolute URL of the metadata document."""
        parts = urlsplit(self.resource)
        return urlunsplit((parts.scheme, parts.netloc, self.metadata_path(), "", ""))

    def challenge(self, error=None, error_description=None):
        """Build the ``WWW-Authenticate: Bearer …`` header value (RFC 6750 + RFC 9728 §5.1), always pointing
        at this resource's metadata."""
        params = []
        if error:
            params.append(f'error="{error}"')
            if error_description:
                params.append(f'error_description="{_quote(error_description)}"')
        params.append(f'resource_metadata="{self.metadata_url()}"')
        return "Bearer " + ", ".join(params)

    def authenticate(self, authorization_header, required_scopes=None):
        """Guard a request: extract the bearer token, validate it (audience must be this resource, plus any
        ``required_scopes``), and return an :class:`AuthDecision`. Missing/invalid/expired/wrong-audience →
        401; insufficient scope → 403."""
        token = bearer_token(authorization_header)
        if token is None:
            return AuthDecision(False, status=401, www_authenticate=self.challenge(),
                                error_description="authentication required")
        result = self.validator.validate(token, required_scopes)
        if result.valid:
            return AuthDecision(True, result=result)
        status, oauth_error = _http_error(result.error)
        return AuthDecision(False, result=result, status=status, error=oauth_error,
                            error_description=result.error_description,
                            www_authenticate=self.challenge(oauth_error, result.error_description))


def _quote(value):
    return str(value).replace("\\", "\\\\").replace('"', '\\"')

File: typescript/src/validator/resource.ts
/**
 * OAuth 2.0/2.1 protected-resource conventions on top of {@link AccessTokenValidator}.
 *
 * A {@link ProtectedResource} is the resource-server side of any OAuth-protected HTTP service — an MCP
 * server (OAuth 2.1 resource server per the MCP authorization spec), an A2A agent, or a plain REST API. It
 * adds what the validator itself doesn't: RFC 9728 Protected Resource Metadata
 * (`/.well-known/oauth-protected-resource`), RFC 6750 bearer extraction + `WWW-Authenticate` challenges, and
 * a request guard that binds the token audience to this resource (RFC 8707). Protocol-neutral — not MCP-only.
 */
import * as errors from "./errors.js";
import type { ValidationError } from "./errors.js";
import type { ValidationResult } from "./result.js";
import type { AccessTokenValidator } from "./validator.js";

const WELL_KNOWN = "/.well-known/oauth-protected-resource";

/** Extract the token from an `Authorization: Bearer <token>` header value, or null. */
export function bearerToken(authorizationHeader?: string | null): string | null {
  if (!authorizationHeader) return null;
  const match = /^\s*Bearer\s+(.+?)\s*$/i.exec(authorizationHeader);
  return match ? match[1] : null;
}

/** RFC 9728 protected-resource metadata document. */
export interface ProtectedResourceMetadata {
  resource: string;
  authorization_servers: string[];
  bearer_methods_supported: string[];
  scopes_supported?: string[];
}

/** The outcome of guarding a request. */
export interface AuthDecision {
  authorized: boolean;
  result?: ValidationResult;
  status: number;
  wwwAuthenticate?: string;
  error?: string;
  errorDescription?: string | null;
}

/**
 * An OAuth-protected resource server. The `validator` should be configured so that `resource` is an
 * accepted audience — that binds incoming tokens to this resource (RFC 8707), which MCP servers MUST enforce.
 */
export class ProtectedResource {
  readonly resource: string;
  readonly authorizationServers: string[];
  readonly scopesSupported: string[];
  readonly #validator: AccessTokenValidator;

  constructor(
    resource: string,
    authorizationServers: string[],
    validator: AccessTokenValidator,
    scopesSupported: string[] = [],
  ) {
    this.resource = resource;
    this.authorizationServers = authorizationServers;
    this.#validator = validator;
    this.scopesSupported = scopesSupported;
  }

  metadata(): ProtectedResourceMetadata {
    const md: ProtectedResourceMetadata = {
      resource: this.resource,
      authorization_servers: this.authorizationServers,
      bearer_methods_supported: ["header"],
    };
    if (this.scopesSupported.length > 0) md.scopes_supported = this.scopesSupported;
    return md;
  }

  metadataPath(): string {
    const path = new URL(this.resource).pathname;
    return path && path !== "/" ? WELL_KNOWN + path : WELL_KNOWN;
  }

  metadataUrl(): string {
    const u = new URL(this.resource);
    return `${u.protocol}//${u.host}${this.metadataPath()}`;
  }

  challenge(error?: string, errorDescription?: string | null): string {
    const params: string[] = [];
    if (error) {
      params.push(`error="${error}"`);
      if (errorDescription) params.push(`error_description="${quote(errorDescription)}"`);
    }
    params.push(`resource_metadata="${this.metadataUrl()}"`);
    return "Bearer " + params.join(", ");
  }

  async authenticate(authorizationHeader?: string | null, requiredScopes?: string[]): Promise<AuthDecision> {
    const token = bearerToken(authorizationHeader);
    if (token == null) {
      return { authorized: false, status: 401, wwwAuthenticate: this.challenge(), errorDescription: "authentication required" };
    }
    const result = await this.#validator.validate(token, requiredScopes);
    if (result.valid) {
      return { authorized: true, result, status: 200 };
    }
    const [status, oauthError] = httpError(result.error);
    return {
      authorized: false,
      result,
      status,
      error: oauthError,
      errorDescription: result.errorDescription,
      wwwAuthenticate: this.challenge(oauthError, result.errorDescription),
    };
  }
}

function httpError(error: ValidationError | null): [number, string] {
  return error === errors.INSUFFICIENT_SCOPE ? [403, "insufficient_scope"] : [401, "invalid_token"];
}

function quote(value: string): string {
  return value.replace(/\\/g, "\\\\").replace(/"/g, '\\"');
}
